shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/BlockChain.hs

```diff
@@ -850,3 +855,26 @@ incrBlocks isOverlay hk b'@(BlocksMade b)
     Just n -> Map.insert hk (n + 1) b
   where
     hkVal = Map.lookup hk b
+
+-- | The validity of any individual block depends only on a subset
+-- of the UTxO stored in the ledger state. This function returns
+-- the transaction inputs corresponding to the required UTxO for a
+-- given Block.
+--
+-- This function will be used by the consensus layer to enable storing
+-- the UTxO on disk. In particular, given a block, the consensus layer
+-- will use 'neededTxInsForBlock' to retrived the needed UTxO from disk
+-- and present only those to the ledger.
+neededTxInsForBlock ::
+  ( Era era,
+    HasField "inputs" (Core.TxBody era) (Set (TxIn (Crypto era)))
+  ) =>
+  Block era ->
+  Set (TxIn (Crypto era))
+neededTxInsForBlock (Block' _ txsSeq _) = Set.filter isNotNewInput allTxIns
+  where
+    getTxIn = getField @"inputs" . getField @"body"
+    txs = Era.fromTxSeq txsSeq
+    allTxIns = foldl' (\acc tx -> acc `Set.union` (getTxIn tx)) mempty txs
+    newTxIds = map (txid . getField @"body") (toList txs)
+    isNotNewInput (TxIn txID _) = txID `notElem` newTxIds
```

Comment:
One thing that is bad here is that `notElem` O(n), instead of `Set.notMember` which O(log n), otherwise seems good. At the same time I think rewriting it this way is a bit clearer:

```suggestion
    txBodies = map (getField @"body") $ toList $ Era.fromTxSeq txsSeq
    allTxIns = Set.unions $ map txins txBodies
    newTxIds = Set.fromList $ map txid txBodies
    isNotNewInput (TxIn txID _) = txID `Set.notMember` newTxIds
```